libzeth/circuits/mimc/mimc_input_hasher.tcc

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2015-2021 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZETH_CIRCUITS_MIMC_MIMC_INPUT_HASHER_TCC__
+#define __ZETH_CIRCUITS_MIMC_MIMC_INPUT_HASHER_TCC__
+
+#include "mimc_input_hasher.hpp"
+
+namespace libzeth
+{
+
+template<typename FieldT, typename comp_fnT>
+mimc_input_hasher<FieldT, comp_fnT>::mimc_input_hasher(
+    libsnark::protoboard<FieldT> &pb,
+    const libsnark::pb_variable_array<FieldT> inputs,
+    const libsnark::pb_variable<FieldT> result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix), _result(result)
+{
+    const size_t num_inputs = inputs.size();
+    if (num_inputs < 2) {
+        // Although it would be superfluous, we could support 1 entry. However
+        // it would add some complexity to the code below. For now, assume
+        // strictly more than 1 entry.
+        throw std::invalid_argument(
+            "inputs array must have at least 2 entries");
+    }
+
+    // Require one compression function invocation per element in the array,
+    // followed by the finalization step. Each invocation except the last
+    // requires an intermediate output variable.
+    _compression_functions.reserve(num_inputs + 1);
+    _intermediate_values.allocate(
+        pb, num_inputs, FMT(annotation_prefix, "intermediate_values"));
+
+    libsnark::pb_linear_combination<FieldT> iv;
+    iv.assign(pb, get_iv());
+
+    // First step: hash_output[0] <- mimc_mp(iv, i[0])
+
+    _compression_functions.emplace_back(new comp_fnT(
+        pb,
+        iv,
+        inputs[0],
+        _intermediate_values[0],
+        FMT(annotation_prefix, " compression_functions[0]")));
+
+    // Intermediate invocations of the compression fucntion.
+    for (size_t i = 1; i < num_inputs; ++i) {
+        _compression_functions.emplace_back(new comp_fnT(
+            pb,
+            _intermediate_values[i - 1],
+            inputs[i],
+            _intermediate_values[i],
+            FMT(annotation_prefix, " compression_functions[%zu]", i)));
+    }
+
+    // Last invocation of compression function to finalize.
+    libsnark::pb_linear_combination<FieldT> num_inputs_lc;
+    num_inputs_lc.assign(pb, FieldT(num_inputs));
+    _compression_functions.emplace_back(new comp_fnT(
+        pb,
+        _intermediate_values[num_inputs - 1],
+        num_inputs_lc,
+        result,
+        FMT(annotation_prefix, " compression_functions[%zu]", num_inputs)));
+
+    assert(_compression_functions.size() == num_inputs + 1);
+}
+
+template<typename FieldT, typename comp_fnT>
+void mimc_input_hasher<FieldT, comp_fnT>::generate_r1cs_constraints()
```

Comment:
`comp_fnT` is not in line with the coding standards: https://github.com/clearmatics/zeth/blob/develop/CODING_STANDARDS.md#naming-and-ordering-conventions - let's switch to `CompFnT` or `compFnT`  instead

Reply:
renamed the type param

Reply:
Thanks